### PR TITLE
core: Parse `module` correctly when `allow-unregistered-dialect` is set

### LINKED ISSUE #3668: core: `module{}` parsed as an unregistered operation when `MLContext(allow_unregistered=True)`
`module {}`, which is a valid spelling of builtin.module {} in MLIR, is parsed as an unregistered operation, even though it's always registered. This happens in particular when `MLContext(allow_unregistered=True)` instead of it's default, which is False.

### PR DESCRIPTION
**Pull request by alexarice**
Fixes #3668

### CODE REVIEW

**Review by dshaaban01 — APPROVED** [review]
Amazing🥳 thank you!

**Review by superlopuh — COMMENTED** [review]
This LGTM, but I'm wondering whether we could simplify things by passing the dialect stack to `get_optional_op` directly, avoiding the creation of extra ops?

**Comment by alexarice** [comment]
> This LGTM, but I'm wondering whether we could simplify things by passing the dialect stack to `get_optional_op` directly, avoiding the creation of extra ops?

Sounds good, I knew there was something I didn't like about my solution

**Comment by alexarice** [comment]
I've rewritten it with @superlopuh suggestion. This didn't quite fix the needless creation of unregistered ops so I've moved some things to a helper function. Also added 2 unit tests.

**Review by math-fehr — APPROVED** [review]
Nice!

File: tests/test_mlcontext.py [hunk 1]
```diff
@@ -60,6 +60,40 @@ def test_get_op_unregistered():
...
+    """
+    Test `get_op` and `get_optional_op`
+    methods with the `allow_unregistered` flag.
+    """
+    ctx = MLContext(allow_unregistered=True)
+    ctx.load_op(DummyOp)
+
+    assert ctx.get_optional_op("dummy", dialect_stack=("test",)) == DummyOp
+    op_type = ctx.get_optional_op("dummy2", dialect_stack=("test",))
+    print(op_type)
```
Comment (superlopuh) [inline tests/test_mlcontext.py:86]: Could you please open a mini PR removing this?

Comment (alexarice) [reply tests/test_mlcontext.py:86]: #3673

**Pull request merged**